hedera-node/src/main/java/com/hedera/services/ledger/interceptors/StakeAwareAccountsCommitsInterceptor.java

```diff
@@ -0,0 +1,319 @@
+				payRewardIfRewardable(pendingChanges, exStakeeI, hasBeenRewarded, latestEligibleStart);
+				// do we need this ?? since stakedToMe is only for account stakes
+			}
+		}
+		if (newStakeeNum > 0) {
+			// Add pending balance to new stakee
+			final var newStakeeI = findOrAdd(newStakeeNum, pendingChanges);
+			manager.updateStakedToMe(newStakeeI, finalBalanceGiven(account, changes), pendingChanges);
+			if (newStakeeI == changesSize) {
+				changesSize++;
+			} else if (!hasBeenRewarded.contains(newStakeeNum)) {
+				payRewardIfRewardable(pendingChanges, newStakeeI, hasBeenRewarded, latestEligibleStart);
+				// do we need this ?? since stakedToMe is only for account stakes
+			}
+		}
+		return changesSize;
+	}
+
+	void payRewardIfRewardable(
+			final EntityChangeSet<AccountID, MerkleAccount, AccountProperty> pendingChanges,
+			final int stakeeI,
+			final Set<Long> hasBeenRewarded,
+			final long latestEligibleStart) {
+		final var account = pendingChanges.entity(stakeeI);
+		final var changes = pendingChanges.changes(stakeeI);
+		final var accountNum = pendingChanges.id(stakeeI).getAccountNum();
+		if (isRewardable(account, changes, latestEligibleStart)) {
+			payReward(account, changes, accountNum, hasBeenRewarded);
+		}
+	}
+
+	private void payReward(final MerkleAccount account,
+			final Map<AccountProperty, Object> changes,
+			final long accountNum,
+			final Set<Long> hasBeenRewarded) {
+		final var reward = rewardCalculator.updateRewardChanges(account, changes);
+		sideEffectsTracker.trackRewardPayment(accountNum, reward);
+		hasBeenRewarded.add(accountNum);
+	}
+
+	int findOrAdd(
+			final long accountNum,
+			final EntityChangeSet<AccountID, MerkleAccount, AccountProperty> pendingChanges
+	) {
+		final var n = pendingChanges.size();
+		for (int i = 0; i < n; i++) {
+			if (pendingChanges.id(i).getAccountNum() == accountNum) {
+				return i;
+			}
+		}
+		// This account wasn't in the current change set
+		pendingChanges.include(
+				STATIC_PROPERTIES.scopedAccountWith(accountNum),
+				accounts.get().get(EntityNum.fromLong(accountNum)),
+				Map.of());
+		return n;
+	}
+
+	boolean isRewardable(
+			@Nullable final MerkleAccount account,
+			@NotNull final Map<AccountProperty, Object> changes,
+			final long latestRewardableStakePeriodStart
+	) {
+		Boolean changedDecline = (Boolean) changes.get(DECLINE_REWARD);
+		return account != null
+				&& account.getStakedId() < 0
+				&& networkCtx.get().areRewardsActivated()
+				&& hasStakeFieldChanges(changes)
+				&& isWithinRange(account.getStakePeriodStart(), latestRewardableStakePeriodStart)
+				&& !Boolean.TRUE.equals(changedDecline)
+				&& (!account.isDeclinedReward() || Boolean.FALSE.equals(changedDecline));
+	}
+
+	void activateRewardsIfValid() {
+		if (!shouldActivateStakingRewards()) {
+			return;
+		}
+		networkCtx.get().setStakingRewards(true);
+		stakingInfo.get().forEach((entityNum, info) -> info.clearRewardSumHistory());
```

Comment:
As mentioned above would like to encapsulate such operations in a `StakingInfoManager`.

Reply:
Moved to `StakingInfoManager`